### PR TITLE
[Toast] Fix timer issues

### PR DESCRIPTION
**Pull request by andy-hook**
Address timer issues mentioned in https://github.com/radix-ui/primitives/pull/1669#issuecomment-1252282997

The underlying issue is that the pause handling doesn't account for multiple calls in succession. Because we're now exposing `onPause` and `onResume` it made more sense to address the issue at source with the duplicate events being fired.

@benoitgrelard might be easier to hop on a call to go over this.

### CODE REVIEW

File: packages/react/toast/src/Toast.tsx [hunk 1]
```diff
@@ -173,36 +173,70 @@ const ToastViewport = React.forwardRef<ToastViewportElement, ToastViewportProps>
     React.useEffect(() => {
       const wrapper = wrapperRef.current;
       const viewport = ref.current;
-      if (wrapper && viewport) {
+      if (hasToasts && wrapper && viewport) {
```
Comment (andy-hook) [inline packages/react/toast/src/Toast.tsx:176]: afaik these handlers shouldn't be necessary when there are no toasts.

File: packages/react/toast/src/Toast.tsx [hunk 2]
```diff
@@ -173,36 +173,70 @@ const ToastViewport = React.forwardRef<ToastViewportElement, ToastViewportProps>
...
         const handlePause = () => {
-          const pauseEvent = new CustomEvent(VIEWPORT_PAUSE);
-          viewport.dispatchEvent(pauseEvent);
-          context.isClosePausedRef.current = true;
+          if (!context.isClosePausedRef.current) {
```
Comment (andy-hook) [inline packages/react/toast/src/Toast.tsx:178]: If we're already paused / running then we shouldn't be dispatching duplicate events.

File: packages/react/toast/src/Toast.tsx [hunk 3]
```diff
@@ -173,36 +173,70 @@ const ToastViewport = React.forwardRef<ToastViewportElement, ToastViewportProps>
...
+          }
+        };
+
+        const isFocusProxy = (element: HTMLElement | null) => {
+          const tailProxyNode = tailFocusProxyRef.current;
+          const headProxyNode = headFocusProxyRef.current;
+          return element === tailProxyNode || element === headProxyNode;
+        };
+
+        const handleFocusOutResume = (event: FocusEvent) => {
```
Comment (andy-hook) [inline packages/react/toast/src/Toast.tsx:199]: When focusing away from `window` we were seeing duplicate events as there was an overlap between `focusout` and `window.blur`, they both fire resulting in `onPause` -> `onResume` -> `onPause`.

As a consumer I wouldn't expect to see this, if they are already paused then the handlers shouldn't fire until that perceivably changes.

File: packages/react/toast/src/Toast.tsx [hunk 4]
```diff
@@ -173,36 +173,47 @@ const ToastViewport = React.forwardRef<ToastViewportElement, ToastViewportProps>
...
+          if (context.isClosePausedRef.current) {
+            const resumeEvent = new CustomEvent(VIEWPORT_RESUME);
+            viewport.dispatchEvent(resumeEvent);
+            context.isClosePausedRef.current = false;
+          }
+        };
+
+        const handleFocusOutResume = (event: FocusEvent) => {
+          // Only resume when focus moves out of the wrapper
+          if (!wrapper.contains(event.relatedTarget as HTMLElement)) {
```
Comment (andy-hook) [inline packages/react/toast/src/Toast.tsx:195]: We still need to handle for this as `focusout / focusin` will fire while tabbing within wrapper and we don't want to flip the flag until we're back outside.

**Comment by benoitgrelard** [comment]
I think I spotted another issue:

https://user-images.githubusercontent.com/1539897/192280746-b25a2aa5-784b-421e-b487-3ca549fee40b.mp4

Leave should only trigger resume if focus isn't inside right?

**Comment by andy-hook** [comment]
> Leave should only trigger resume if focus isn't inside right?

Good spot, tweaked a little. Let me know how that feels.

**Review by benoitgrelard — APPROVED** [review]

**Pull request merged**